### ISSUE TITLE
Nodes should know peers by name

### ISSUE THREAD

**Issue by yfeldblum** [edelight/chef-mongodb#2]
There should be an option for which node attribute to pull as the name by which peer nodes are known. For example, instead of one node knowing about its peers by their `peer['ipaddress']`, one might like to use `peer['fqdn']`, `peer['name']`, `peer['private_ipaddress']`, `peer['cloud']['local_ipv4']`, etc. If anything, nodes should know each other by `peer['fqdn']`, because FQDNs can be translated to ipaddresses by `/etc/hosts` or the dns.

**Comment by thekorn** [comment]
Hi yfeldblum,
thanks for reporting this issue. I agree with you, peers should really know each other by fqdn rather than by ipaddress.
But I don't think making the identifier configurable makes real sense, as this will make the mongodb cookbook much harder to read.
I will change ipaddress to fqdn in my next commit.
Would this compromise work for you?

Markus

**Comment by yfeldblum** [comment]
I would say that's the best overall approach.

I would also suggest mentioning in the readme that this is now the case, and that consumers of the cookbook _must_ set their `/etc/hosts` file or set up host-based DNS (such as `dnsmasq`) or each host or set up a more centralized DNS server/cluster. All of which should be done anyway. Unless the consumer is running on EC2 which has this built in.

**Comment by mattray** [comment]
Using fqdn isn't going to work for rackspace-based servers (not sure about EC2 and other cloud providers). Their fqdns come back as 'slice20226102' for example, which can't be resolved by other rackspace boxes.

**Comment by yfeldblum** [comment]
The `ipaddress` key isn't going to work for Rackspace Cloud-based servers either, because that's the Internet interface, which is slow, billed, and susceptible to snooping/spoofing attacks, rather than the LAN interface, which at Rackspace Cloud is fast, free, and protected from snooping/spoofing attacks.

For Rackspace Cloud-based servers, we either require to know _which_ attribute, however many levels deep, we need; or we need to assume that the sysadmin can set up the hosts file, a local DNS resolver on each host, or a central DNS server.

```
# recipe
ipaddress_attribute = Array(node["mongodb"]["ipaddress_attribute"])
ipaddress = ipaddress_attribute.inject({}) { |attr, key| attr[key] }

# role or environment
override_attributes \
  "mongodb" => {
    "ipaddress_attribute" => ["cloud", "local_ipv4"]
  }
```

**Comment by thekorn** [comment]
@mattray: is there any alternative which would work for all systems. My understanding of chef was that it needs fqdn attributes of nodes to work correctly.

**Comment by jtimberman** [comment]
'ohai' will detect if it is running on ec2, rackspace or eucalyptus and populate a 'cloud' attribute with sub-attributes. You can check if node['cloud'] exists, and if it does, use node['cloud']['private_ipv4'] as the IP address. We do this in our cookbooks that utilize search, such as the application::rails.

https://github.com/opscode/cookbooks/blob/master/application/recipes/rails.rb#L125

**Comment by tobami** [comment]
@jtimberman: looks like a good compromise. It wouldn't be the same case as the rails example though. The first I though wes that we could just add a new attribute:
`default[:mongodb][:fqdn_attribute] = node.attribute?('cloud') ? node['cloud']['local_ipv4'] : node['fqdn']`

and keep search the same. However, that logic should go inline where the search statements take place, because we are not interested in the "cloud" value for the current node. It is something that will be set independently for each node. One instance of a replicaset could be in EC2, and another in your own data center for example (i.e. no node['cloud'] attributes).

We'll see when we get to implement that, but it would be best if someone using cloud servers would send us a patch. The search statements that need to be changed are in /definitions/mongodb.rb and in  libraries/mongodb.rb (several)

**Comment by yfeldblum** [comment]
That raises the question: how do you do cross-region and cross-provider clusters? The `cloud/local_ipv4` attribute is a **LAN** IP address and will not work cross-region or cross-provider.

How would one set up a replica set with two node in one region and one node in another - for example, two in an east-coast datacenter and one in a west-coast datacenter? That seems like a scenario that would be in demand.

**Comment by tobami** [comment]
Right, that is why I think fqdn is best, and for cross-datacenters you will probably need either custom /etc/hosts or a DNS server, as you suggested.

**Comment by josephholsten** [comment]
Hello everyone from two years ago!

While it drives me nuts that people keep referring to FQDNs that don't have matching PTR records (I'm looking at you rackspace), I do get the appeal of using a domain name instead of a possibly private network IP. Responsible admins use interface aware domain names and DNS. It truly is the Right Way (tm).

That said, plenty of folks are plenty happy with something less than the Right Way. I'd like to see a solution that defaults to an FQDN, but allows node search to pick identifiers in a parameterized way.

As I mentioned in https://github.com/edelight/chef-mongodb/pull/119, I feel like the core definitions or RPs should not assume any awareness of the nodes they're supposed to connect to. I'm almost fond of a nice generalized discovery tool like [hw-cookbooks/discovery](https://github.com/hw-cookbooks/discovery), especially its `Discovery.ipaddress`.

So what's the consensus around here? Would everyone be happy with a default of `node['fqdn']`, but a node attribute so we can either use smart IP address discovery or an explicit node attribute on discovered nodes?

And how do people feel about adding a dependency on heavywater's discovery cookbook?
